### PR TITLE
[masterbots.ai] feat: email verification

### PR DESCRIPTION
**Pull request by Bran18**
### Preview video:

https://github.com/user-attachments/assets/438e69d7-124a-4755-9f8a-daa35c3e7aaf

### Alternatives for the CRON 

``` json
{
  "crons": [
    // Run at midnight UTC
    {
      "path": "/api/cron/handle-unverified-users",
      "schedule": "0 0 * * *"
    },

    // Run twice daily (00:00 and 12:00 UTC)
    {
      "path": "/api/cron/handle-unverified-users",
      "schedule": "0 0,12 * * *"
    },

    // Run every 6 hours
    {
      "path": "/api/cron/handle-unverified-users",
      "schedule": "0 */6 * * *"
    }
  ]
}

```

## Summary by Sourcery

Introduce email verification for new user sign-ups, including sending verification emails and handling verification tokens. Implement a cron job to manage unverified users, sending reminders and deleting accounts after 15 days. Add a new email verification page to guide users through the verification process. Enhance the sign-up form with client-side validation and improve email handling with template generation.

New Features:
- Implement email verification process for new user sign-ups, including sending verification emails and handling verification tokens.
- Add a cron job to handle unverified users, sending reminder emails and deleting accounts after 15 days if not verified.
- Introduce a new email verification page to guide users through the verification process.

Enhancements:
- Refactor the sign-up form to include client-side validation for password matching and display verification notices.
- Enhance email handling by adding a function to generate email templates for verification and password reset.

## Release Notes

- **New Features**
	- Introduced email verification during user signup, enhancing account security.
	- Added a new endpoint for verifying email addresses.
	- Implemented a cron job to manage unverified users, including sending reminders and blocking accounts after 15 days.
	- Added email templates for password reset and verification.
	- Enhanced the SignInPage to display a message upon successful email verification.
	- Created a new email verification page for user experience.

- **Bug Fixes**
	- Improved error handling in password reset and signup processes.

- **Documentation**
	- Updated comments and structure for clarity in the signup and password reset functionalities.

- **Refactor**
	- Enhanced state management in the signup form component.
	- Streamlined logic for user existence checks and token generation.

- **Style**
	- Adjusted UI elements for better user experience in the sign-in and verification pages.

### CODE REVIEW

File: apps/masterbots.ai/components/auth/signup-form.tsx [hunk 1]
```diff
@@ -1,47 +1,111 @@
...
+          <Button
+            variant="outline"
+            onClick={() => router.push('/auth/signin')}
+          >
+            Go to Sign In
+          </Button>
+        </div>
+      </div>
+    )
+  }
```
[the remark this answers is not in the document]
Comment (Bran18) [reply apps/masterbots.ai/components/auth/signup-form.tsx:96]: Good call, but is better not to provide this to the user yet.

File: apps/masterbots.ai/components/auth/signup-form.tsx [hunk 2]
```diff
@@ -1,47 +1,111 @@
...
+        toast.success('Account created successfully! Please check your email to verify your account.')
+      } else {
+        toast.error(data.error || 'Failed to sign up')
+      }
+    } catch (error) {
+      console.error(error)
+      toast.error('An unexpected error occurred')
+    } finally {
+      setState(prev => ({ ...prev, isLoading: false }))
     }
```
[the remark this answers is not in the document]
Comment (Bran18) [reply apps/masterbots.ai/components/auth/signup-form.tsx:67]: I'll include the validations that I created for the Password Recovery in another PR with another improvements and changes

File: apps/masterbots.ai/lib/email.ts [hunk 3]
```diff
@@ -102,3 +108,103 @@ export async function testEmailConfig() {
...
+          </tr>
+        </table>
+      </body>
+      </html>
+    `
+  }
+
+  return templates[type]
+}
+
```
[the remark this answers is not in the document]
Comment (Bran18) [reply apps/masterbots.ai/lib/email.ts:182]: Going to create a util/template to have the emails there in a better way in another PR

[the remark this answers is not in the document]
Comment (Bran18) [reply apps/masterbots.ai/lib/email.ts:182]: Sure, and assigned me to the task

**Review by AndlerRL — APPROVED** [review]
LGTM! Nice approaches to deal with the email verification and confirmation :+1: I would lean though to update the cronjob to update every day at 00:00 UTC after that change, we are good to go!

> **PRO TIP:** When doing a major flow, sometimes we have to do many side-effects (within the API route on this case) in order to complete a whole system. This way how you did is fine though it is fully recommended to be modular with the main flow, so you can clearly see what actions comes first and which action follows next easier. Making code comments also helps and their length are good, reason why it will not be required a modification.

**Comment by Bran18** [comment]
> lean though to update the cronjob to update every day at 00:00 UTC

CRON added to expression 0 0 * * * that means:

```0 - At minute 0
0 - At hour 0 (midnight)
* - Every day
* - Every month
* - Every day of the week
```

Thank you very much for the PRO TIP. I'll 100% think about improving our AUTH APIS and actions :pray:

**Comment by Bran18** [comment]
@AndlerRL good to go, i just added the `is_blocked` to true instead deleting xD

**Pull request merged**